drivers: sensor: Fix TMAG5273/TMAG3001 by adding a TMAG3001 compatible

yiding opened the pull request:
TMAG5273 and TMAG3001 share much of the similar functionality.

The previous code attempted to detect whether TMAG5273 or TMAG3001 was connected based on DEVICE ID register. This doesn't work as the bits that denote the version on one part are undefined on the other part, and cannot be relied on to be zero.

This commit adds a TMAG3001 compatible which (for now) is just derived from the TMAG5273 compatible. This allows TMAG3001 to be specified directly in the DT. The driver code is updated to support both compatibles.

I thought about moving the common properties to a `ti,tmagxxxx.yaml`, however there's other parts (namely, tmag5170) which is SPI and have different properties, so I did not do this. We might still want to pull out some common properties, as TMAG5273 does have functionality that doesn't exist in TMAG3001, which currently can't be reflected in the device tree yaml.

## Test plan

I tested this with TMAG5273, but do not have a TMAG3001. However, this code shouldn't change any of that functionality. I did try the tmag3001 compatible to see that it builds.

yiding commented:
There's some whitespace changes in tmag5273.h file. Seems like clang-format changed tabs to spaces in alignment spaces for the defines. Not sure if that's expected or my clang-format is somehow misconfigured (it does use leading tabs for indentation).

gumulka commented:
> There's some whitespace changes in tmag5273.h file. Seems like clang-format changed tabs to spaces in alignment spaces for the defines. Not sure if that's expected or my clang-format is somehow misconfigured (it does use leading tabs for indentation).

The file was previously not correctly formatted. But you should try to keep the old wrong formatting, so that the change is not that big and one can see what has changed in your commit. Maybe add a second pull request for the formatting.

But thanks for the pull request. It was on my plate for a long time and I never came around to actually doing it.

yiding commented:
Reverted formatting changes.

MaureenHelm reviewed (commented):
Add the new compatible to `tests/drivers/build_all/sensor/i2c.dtsi`

yiding commented:
Added the sensor to the build test. Also, that test is quite the merge conflict factory with all sensors being added at the end.

MaureenHelm commented:
> Added the sensor to the build test. Also, that test is quite the merge conflict factory with all sensors being added at the end... might be cool if it's not one giant hand-edited file, like if the example declaration can be in a dtsi for each sensor driver, and some script comes and collects them all.

PRs are welcome

kartben commented:
@yiding can you please rebase again? Thanks

MaureenHelm commented on tests/drivers/build_all/sensor/i2c.dtsi:
```diff
@@ -1185,3 +1185,13 @@ test_i2c_lsm6dso32: lsm6dso32@a5 {
 	gyro-odr = <LSM6DSO_DT_ODR_6667Hz>;
 };
+
+test_i2c_tmag3001: tmag3001@9d {
```
```suggestion
test_i2c_tmag3001: tmag3001@a6 {
```

yiding committed:
drivers: sensor: Fix TMAG5273/TMAG3001 by adding a TMAG3001 compatible

Fixes issue introduced in #76460

The previous code attempted to detect whether TMAG5273 or TMAG3001 was
connected based on DEVICE ID register. This doesn't work as the bits that
denote the version on one part are undefined on the other part, and cannot
be relied on to be zero.

This commit adds a TMAG3001 compatible which (for now) is just derived from
the TMAG5273 compatible. This allows TMAG3001 to be specified directly in
the DT. The driver code is updated to support both compatibles.

Signed-off-by: Yiding Jia <yiding.jia@gmail.com>

yiding commented:
Feels like we are stuck in an endless loop here.
